internal/auth/oidc/repository_auth_method_update.go

```diff
@@ -1,14 +1,28 @@
 package oidc
 
-import "context"
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/hashicorp/boundary/internal/db"
+	dbcommon "github.com/hashicorp/boundary/internal/db/common"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+	"github.com/hashicorp/boundary/internal/oplog"
+	"github.com/hashicorp/boundary/sdk/strutil"
+	"github.com/hashicorp/go-multierror"
+)
 
 // UpdateAuthMethod will retrieve the auth method from the repository,
 // update it based on the field masks provided, and then validate it using
 // Repository.TestAuthMethod(...).  If the test succeeds, the auth method
 // is persisted in the repository and the written auth method is returned.
 // fieldMaskPaths provides field_mask.proto paths for fields that should
 // be updated.  Fields will be set to NULL if the field is a
-// zero value and included in fieldMask. Name, Description, State, DiscoveryUrl,
+// zero value and included in fieldMask. Name, Description, OperationalState, DiscoveryUrl,
```

Comment:
line #100 calls ValidateAuthMethod(..) to the applied changes request in the update.  If that fails, an error is returned and the changes are not persisted. 

Changing other values does not have any side-affects on the Auth Method's state.  For instance, if the caller sets the SigningAlg to ES52 (which is not published in the method's discovery info) on an AuthMethod that's ActivePublic, and provides the WithForce() option, the update will update the DiscoveryUrl and have no affect on the state which will remain ActivePublic.    This is by design and intentional.  The API needs to support this use-case.

Reply:
We need to check the validity of an AuthMethod for several operations: Updates and OperationalState changes.  At this point, I'm not sure why we would force a caller to make two calls when they want to update an AuthMethod's values and then it's OperationalState, since we need to check the validity of the AuthMethod for both operations.

I do see some value in providing explicit OperationalState change endpoints... and I've implemented these narrowly defined update operations in the jimlambrt-oidc-wip branch for a the next PR.  They still need unit tests.

Reply:
Hum.. I just noticed that the RFC doesn't specify a way to transition to ActivePublic or ActivePrivate unless the Auth Method is successfully validated against the IdP's discovery document.  I think, that's a poor choice.  We should allow some sort of WithForce(...) option for these operations as well.  If the operator is sure the config will work and the IdP's discovery info is out of date, they have to have a way to enable the auth method for users.

Reply:
>There is custom actions dedicated to /v1/auth-methods/amoidc_1234567890:change-state:make-inactive, /v1/auth-methods/amoidc_1234567890:change-state:make-private, and /v1/auth-methods/amoidc_1234567890:change-state:make-public

I realize most people will use SDK or UI, but -- doesn't this seem overly wordy? Why not just `change-state` with the desired state as a parameter?

Reply:
@jefferai @talanknight - if there are no objections, I'm going to add a TODO to `UpdateAuthMethod` about the need to align on supporting the update of OperationalState or not, so I can proceed with a merge.

Reply:
Sure - my question isn't really about the code here, it's about the API aspects, so that's fine by me.

Reply:
In the RFC, the `OperationalState` is calculated and set by the system not a user. This is a usability feature. It prevents end users from seeing or selecting an AuthMethod that is provably misconfigured which in turn protects an administrator from exposing an error and sparing them some potential embarrassment.

The calculation happens when any of the non authentication operations are called. The calculation uses the configuration of the AuthMethod and the operation invoked to determine the OperationalState. An AuthMethod that has a provably wrong configuration is always in the inactive state. 

I believe a configuration missing any of the following items would qualify as provably wrong:
- discovery url
- client id
- client secret
- callback url
- signing algorithm

Reply:
see :https://github.com/hashicorp/boundary/pull/963/commits/2b110c144f910dee53f6d9e0b4bb091a11102b83